pkgs/coverage/test/isolate_paused_listener_test.dart

```
@@ -889,5 +894,40 @@ void main() {
         // Don't try to resume B, because the VM service is already shut down.
       ]);
     });
+
+    test('throw when resuming main isolate is ignored', () async {
+      resumeFailures = {'main'};
+
+      startEvent('main', '1');
+      startEvent('other', '2');
+      pauseEvent('other', '2');
+      exitEvent('other', '2');
+      pauseEvent('main', '1');
+      exitEvent('main', '1');
+
+      await endTest();
+
+      expect(received, [
+        'Pause other. Collect group 2? Yes',
+        'Resume other',
+        'Pause main. Collect group 1? Yes',
+        'Resume main',
+      ]);
+    });
+
+    test('throw when resuming other isolate is not ignored', () async {
+      resumeFailures = {'other'};
+
+      startEvent('main', '1');
+      startEvent('other', '2');
+      pauseEvent('other', '2');
+      exitEvent('other', '2');
+      pauseEvent('main', '1');
+      exitEvent('main', '1');
+
+      // TODO: Not sure how to write this test, since the RPCError is thrown in
+      // an async handler, and not propagated to this expectation.
+      // expect(() => endTest(), throwsA(isA<RPCError>()));
```

@bkonyi Any suggestions about how to write this test? This expectation doesn't work. The `endTest()` future actually completes successfully. The RPCError is printed to the console as an unhandled exception.

```
00:00 +22 -1: IsolatePausedListener throw when resuming other isolate is not ignored [E]                                  
  resume: (-32000) Application error
  test/isolate_paused_listener_test.dart 533:11             main.<fn>.<fn>.<fn>
  package:mockito/src/mock.dart 186:47                      Mock.noSuchMethod
  test/collect_coverage_mock_test.mocks.dart 1642:14        MockVmService.resume
  package:coverage/src/isolate_paused_listener.dart 108:26  IsolatePausedListener._onPause
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 177:7   listenToIsolateLifecycleEvents.onPause
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 200:16  listenToIsolateLifecycleEvents.<fn>
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 277:7   IsolateEventBuffer.add
  
  resume: (-32000) Application error
  test/isolate_paused_listener_test.dart 533:11             main.<fn>.<fn>.<fn>
  package:mockito/src/mock.dart 186:47                      Mock.noSuchMethod
  test/collect_coverage_mock_test.mocks.dart 1642:14        MockVmService.resume
  package:coverage/src/isolate_paused_listener.dart 108:26  IsolatePausedListener._onPause
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 177:7   listenToIsolateLifecycleEvents.onPause
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 200:16  listenToIsolateLifecycleEvents.<fn>
  ===== asynchronous gap ===========================
  package:coverage/src/isolate_paused_listener.dart 277:7   IsolateEventBuffer.add
  
  Expected: throws <Instance of 'RPCError'>
    Actual: <Closure: () => Future<void>>
     Which: returned a Future that emitted <null>
```

---

`expectLater` doesn't work either. The problem is that the `Future` returned by `endTest`/`waitUntilAllExited` does complete successfully. The exception doesn't propagate through this future. The stack trace terminates at `IsolateEventBuffer.add`, which is invoked by `VmService.onIsolateEvent.listen`.

I thought you might have had to solve this problem in `VmService`'s event listener tests. I feel like there's probably some solution involving `Zone`s, but I can't figure out how to make it work.

---

When running the tool normally, this error should be reported (with the normal stack trace etc), not ignored. Leaving it as an unhandled exception does that. But for this test I want to catch the error and verify it is what I expect. I guess I'll have another go at using a `Zone`.

---

This infra is all about managing the lifecycle of isolates and the VM under test, and the sequencing is pretty complex. This error indicates a bug in that sequencing that could lead to missing coverage (ie failing to collect coverage for an isolate group). If we report the error then it gives me a chance to debug it (or for users to file a bug), but if we ignore the error then the failure mode is that occasionally the coverage report will be incomplete. Much harder to notice and debug.

If I get a report of this error and I'm able to repro it, but it's not clear what's going on and there's no way of fixing the error, and it's rare and harmless, then I'll start ignoring this error. That's what happened for the main isolate error case.